felix/dataplane/linux/endpoint_mgr.go

```diff
@@ -669,6 +669,14 @@ func (m *endpointManager) resolveWorkloadEndpoints() {
 	}
 
 	removeActiveWorkload := func(logCxt *log.Entry, oldWorkload *proto.WorkloadEndpoint, id types.WorkloadEndpointID) {
+		// QoS state must be removed before the workload itself is removed
+		if oldWorkload != nil {
+			logCxt.Info("deleting QoS bandwidth state if present")
+			err := m.maybeUpdateQoSBandwidth(oldWorkload, nil)
+			if err != nil {
+				logCxt.WithField("workload", oldWorkload).Infof("error deleting QoS bandwidth state, workload may have been already removed: %v", err)
```

Comment:
If the CNI plugin deletes the cali interface, does that clean up the associated mirror interface?  If not, how does that eventually get cleaned up?

Reply:
The code that removes the QoS config does not abort when the "main" interface was already removed, so the mirror interface is correctly torn down even if the CNI plugin already removed the "main" interface